README.md
# QMAI-workflow

聚合了公司前端开发工作中的常用功能。便捷大家的操作

## repository

[仓库地址](https://git.zmcms.cn/FrontEnd/qmai-workflow.git)

## idea

大家有什么好的 idea，可以在项目中的[议题](https://git.zmcms.cn/FrontEnd/qmai-workflow/issues/new)中发起。  
大家共同来评审该 idea 合理性，以及排期。

## contribution

大家可以自由认领 task，不可重复认领。  
以提 pr 的形式的来提交代码，会有人员来 review code，讨论并修改，完善后合并入 master 分支。


CHANGELOG.md
# Change Log

此项目的所有显着更改都将记录在此文件中。

格式基于[Keep a Changelog](https://keepachangelog.com/en/1.0.0/)，
并且该项目遵循 [语义版本控制](https://semver.org/spec/v2.0.0.html)。

## [Unreleased]

### Changed

- 所有依赖项更新到当前最新。
- 添加 prettierrc.js 配置文件。

### Removed

- 移除右上方的刷新按钮。

## [1.8.4] - 2022-11-18

### Removed

- 删除**qm-form**的语法高亮。
- **编程导航**移除原项目管理工具**TAPD**的链接。

## [1.8.3] - 2022-11-08

### Fixup

- 修复`vscode@1.73.0`版本以上，插件打开失败。

### Changed

- 修正**快捷导航**模块中的文案错误。
- 发布通知由钉钉转为飞书通知机制。

## [1.8.2] - 2022-09-25

### Added

- **编程导航**模块添加**APP 调试工具**导航入口

### Changed

- 更新**编程导航**模块里面的图标
- 添加更新内容至钉钉群

## [1.8.0] - 2022-09-18

### Changed

- 各个模块添加**SSR_DATA** 加快渲染

### Fixed

- **快捷部署**模块中的**switch**组件按钮的背景色

## [1.7.3] - 2022-09-14

### Fixed

- 修复已选择日志主题的情况下切换日志环境下数据无更新

## [1.7.2] - 2022-09-12

### Changed

- 替换 Qutils 的导航图标

## [1.7.1] - 2022-09-08

### Fixed

- 在 master 分支下使用**一键更新规范** 步骤条不准

## [1.7.0] - 2022-09-08

### Added

- 添加**一键更新规范功能**

### Changed

- 修改**编程导航**中**Qspec**的 icon

### Fixed

- message 组件背景色为空

## [1.6.1] - 2022-09-05

### Fixed

- **部署列表**小型弹窗背景色为空

## [1.6.0] - 2022-09-04

### Added

- **生成规范**模块进一步颗粒化划分'项目类型'、'配置内容'等配制项

### Changed

- 整体项目适配当前的主题颜色
- 优化整体项目的响应式适配
- 优化**生产规范**、**分支记录**、**日志查询**模块表单内容项

## [1.5.0] - 2022-08-24

### Added

- 为 QmForm 提供代码高亮
- **部署列表** 添加搜索栏

## [1.5.0] - 2022-08-24

### Added

- 为 QmForm 提供代码高亮
- **部署列表** 添加搜索栏

## [1.4.3] - 2022-08-03

### Added

- **导航链接** 添加**eslint-plugin-qmai**链接

## [1.4.2] - 2022-07-28

### Added

- **导航链接** 添加**debug**和**云盘**链接

## [1.4.1] - 2022-07-25

### Changed

- **snippets** 再次添加并更新 qmForm 的代码模版 json。

## [1.4.0] - 2022-07-25

### Added

- **编程导航** 添加编程导航模块。

### Fixup

- **sentry** 修复 sentry 的地址链接错误

### Changed

- **生成规范** 更新生成规范的模版。
- **Vite** 接入最新的`vite@3.x` 更快，更好。
- **All** 删除冗余代码于发布的包中，提升安装速度。

### Removed

- **snippets** 删除内置的 qmForm 的代码模版 json。

## [1.3.0] - 2022-07-15

### Added

- **快捷部署** 添加挂起审核功能，无需弹窗强制等待。(#16)
- **快捷部署** 部署失败时候，添加**删除记录**按钮，提供删除该条记录的功能。 (#15)
- **日志查询** 日志主题动态添加权重，为用户记录自身最常用的日志主题。(#5)

### Fixup

- **侧栏** 部分链接错误

### Changed

- **日志查询** 去除日志环境和日志主题的联动效果，增强易用性。
- **快捷部署** 优化大屏幕下，快捷部署表单的宽度。
- **快捷部署** 部署错误的时候去除 ASNI 的转义字符，更为直观的展现错误。
- **快捷部署** 侧栏取消文字选中的颜色 趋向于原生体验

## [1.2.0] - 2022-07-08

### Added

- **快捷部署**部署失败时直接展现错误日志 (#9)
- 添加**Qscode**模块 快速查看插件文档
- 右上角添加刷新按钮 (#13) 感谢 zhouanfeng 同学的 Pr

### Fixup

- 侧栏宽度极小的情况下的样式异常 (#10)
- 不能够正确匹配项目名称大小写差异 (#11)
- 修复**Gitlab**的标签类型为**快捷入口**

### Changed

- 修改**新人必读**模块的展示，修改为浏览器链接

## [1.1.13] - 2022-07-06

### Changed

- 优化**新人阅读**中的文案

## [1.1.12] - 2022-07-04

### Changed

- 优化**生成规范**的 模版文件

## [1.1.11] - 2022-07-04

### Fixup

- 修复**快捷部署**的 Tip 文案
- 修复**生产规范**模块下生成的 spec 的配置错误

## [1.1.10] - 2022-06-30

### Added

- 支持用户 settings.json 来配置 menu icon 的显示与隐藏

## [1.1.9] - 2022-06-29

### Fixup

- 修复存在不规范 tag 的排序错误

## [1.1.8] - 2022-06-29

### Changed

- 重新部署按钮 会再次嗅探本机的环境信息

## [1.1.7] - 2022-06-29

### Fixup

- 降低对 git 版本的依赖

## [1.1.6] - 2022-06-29

### Fixup

- 修复 windows 系统下 fspath 错误

## [1.1.5] - 2022-06-29

### Fixup

- 修复 windows 系统下 siderbar 错误
- 修复亮色主题下 展示不清晰

## [1.1.4] - 2022-06-29

### Added

- 快捷部署模块 添加**faster switch**功能，供用户选择是否更快获取分支信息

## [1.1.3] - 2022-06-28

### Changed

- zx 锁定版本为 6.1.0
- ant-design-vue 升级到 3.2.9

### Fixup

- 发布正式环境下 上线单列表展示问题 => 显示为**prod**而不是**master**
- 未登录前 展现部署列表

## [1.1.2] - 2022-06-28

### Changed

- 默认获取本地项目以及分支信息 加快部署

## [1.1.1] - 2022-06-27

### Added

- 添加**前端食堂技术周刊**链接

### Fixup

- 部署时候 需要审核的业务场景下 强行开始部署的问题

## [1.1.0] - 2022-06-26

### Changed

- 对于整个侧边栏改动 webview 实现
- 添加 pannel 上的图标

## [1.0.5] - 2022-06-22

### Added

- 添加**Qutils**入口

## [1.0.5] - 2022-06-20

### Fixed

- 兼容"部署系统权限 bug"

## [1.0.1] - 2022-06-16

### Changed

- 快捷部署 Tip 文案调整

## [1.0.0] - 2022-06-16

### Changed

- 重新上架市场

## [0.15.2] - 2022-06-16

### Added

- 在右上角菜单栏目添加“快捷部署”的入口

## [0.15.1] - 2022-06-13

### Changed

- 给通过插件进行部署的 title 加上“⭐️”标识来统计使用人数

## [0.15.0] - 2022-06-13

### Added

- 添加“生成规则”模块 来快速生成代码规范配置

## [0.14.0] - 2022-06-11

### Added

- 添加“快捷入口”模块

## [0.13.3] - 2022-06-11

### Changed

- 优化小屏幕下“部署列表”模块下的 searchForm 的样式。

## [0.13.2] - 2022-06-08

### Changed

- 修复“暗色模式”主题下 message 和 Modal 的样式
- create command error

## [0.13.1] - 2022-06-08

### Changed

- 修复“暗色模式”主题下 log 模块的头部背景颜色

## [0.13.0] - 2022-06-08

### Added

- 修改主题为“暗色模式”

## [0.12.0] - 2022-05-30

### Added

- 添加“部署列表”模块

## [0.11.3] - 2022-05-28

### changed

- 修改“新人必读”模块下 windows 默认终端修改方法

## [0.11.2] - 2022-05-28

### changed

- 添加部署系统提示文案
- 修改“新人必读”模块下 windows 默认终端修改方法

## [0.11.1] - 2022-05-24

### changed

- 修复“重启 eslint 插件”按钮 icon 正常

## [0.11.0] - 2022-05-24

### Added

- 编辑器的菜单区添加“重启 eslint 插件”按钮

## [0.10.0] - 2022-05-23

### Added

- 找回“分支记录”模块

## [0.9.2] - 2022-05-19

### Changed

- 新人指南中 Git 模块 文案的更改

## [0.9.1] - 2022-05-15

### Changed

- 新人指南中 Git 模块 文案的更改

## [0.9.0] - 2022-05-15

### Added

- 添加了**新人必读**模块

### Changed

- 上传图片无响应无反馈

### Removed

- 删除了**提交格式**模块

## [0.8.0] - 2022-05-11

### Added

- 添加 qm-form 代码片段

## [0.7.1] - 2022-05-10

### Changed

- 日志系统支持返回登录

## [0.7.0] - 2022-05-09

### Changed

- 更换了`快速部署`功能模块的 cache 路径
- 封装了 useChannel
- 优化各个模块的 ts 相关代码

## [0.6.0] - 2022-05-08

### Added

- 添加`日志查询`功能模块

## [0.5.1] - 2022-05-05

### Removed

- 删除了 oss 默认的开发者的登录账号和密码

## [0.5.0] - 2022-05-05

### Added

- oss 图片上传功能

## [0.4.5] - 2022-05-04

### Changed

- 优化了不处于部署面板时的结果反馈

## [0.4.4] - 2022-05-03

### Changed

- 优化了部署失败场景下的页面反馈
- 优化了不处于部署面板时的结果反馈

## [0.4.3] - 2022-05-01

### Added

- 接入`vitest`进行单元测试 为系统提供保障

## [0.4.2] - 2022-05-01

### Changed

- 再次修改了**常用文档/提交格式**页面展示

## [0.4.1] - 2022-05-01

### Changed

- 修改了**常用文档/提交格式**页面展示

## [0.4.0] - 2022-05-01

### Added

- 添加了**推荐阅读**模块

### Changed

- 移动**Git 管理 / 提交格式**到**常用文档**模块下

### Removed

- 删除了**Git 管理**模块

## [0.3.3] - 2022-04-29

### Changed

- 发布生产环境找寻不到正确的 Tag 版本

## [0.3.2] - 2022-04-27

### Changed

- 上线单的标题时去除表情 来跳过部署系统的缺陷

## [0.3.1] - 2022-04-26

### Added

- 添加插件图标

## [0.3.0] - 2022-04-26

### Added

- 各大文档添加地址

## [0.2.0] - 2022-04-26

### Added

scripts/new-fn.mjs
#!/usr/bin/env zx

import './shared/index.mjs'
import prompts from 'prompts'
import { fs, $ } from 'zx'

// 获取升级后版本号
let { name, cname } = await prompts([
  {
    type: 'text',
    name: 'name',
    message: '功能名称为? (Tip: 英文)',
  },
  {
    type: 'text',
    name: 'cname',
    message: '功能名称为? (Tip: 中文)',
  },
])

// 侧栏添加
const sidebarVueContent = fs.readFileSync('./src/module/sidebar/app.vue', 'utf-8')
fs.writeFileSync(
  './src/module/sidebar/app.vue',
  sidebarVueContent.replace(
    '/** fs-placeholder */',
    `{
    desc: '这里是描述...',
    title: '${cname}',
    tag: ['便捷工具'],
    command: 'qmai.${name}',
    avatar: '<svg viewBox="0 0 125 125"><title>Artboard 25</title><g data-name="person 18"><path data-name="neck" d="M77,80.59s-.67-.29-.67-.88V73l-26,1.29s.29,7.91-.31,8.59c0,0-3.1,2.27-2.2,4.15C49.71,91,55.06,93.35,59.12,94c5,.8,12.63,0,17-2.81C79,89.37,80.84,82.28,77,80.59Z" fill="#fce7d4" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><path data-name="face" d="M81,56.13s-.54,9.19-.73,9.94a12.71,12.71,0,0,1-1.47,3.56,68.55,68.55,0,0,1-5.5,7.52c-1,1-3.11,4.12-4.29,4.46a28.11,28.11,0,0,1-5.85.94c-1.25-.07-5.76-.67-6.75-1.27S52,76.64,49.17,72.91,44.9,66.14,44.7,65.2a42.77,42.77,0,0,1-.26-4.33L44,61s-.06.95-.89,1.06c-1.19.15-1.91-1.35-2.22-2.27-.51-1.53-.29-3.22-.84-4.75s-3.06-6.88-1.06-8.24c1.75-1.19,3.31.49,3.88,2.09a15.67,15.67,0,0,1,.44,1.63l.34-.09s-.05-3,0-5.06c.09-3.88-1.1-8.22.35-12C48.75,21,53.27,20.54,64.38,20.18c9.91-.33,15.47,7.72,16.22,11.4a36.54,36.54,0,0,1,.88,9c-.15,2.73-.52,9.53-.52,9.53l.32.06a28.45,28.45,0,0,1,1.32-3.32,1.72,1.72,0,0,1,3.18,1.29c-.29,1.07-.82,4.14-1,5.38a16.46,16.46,0,0,0,0,2.72c0,.83-.35,3.7-1.57,3.67C82.42,59.85,80.57,59.9,81,56.13Z" fill="#fce7d4" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><path data-name="Sweater" d="M2.84,125.7s.29-1.13,1-4.53,5.22-15,7.61-17.93,6-4.53,8.49-5.74S47.05,83.4,47.05,83.4a29.56,29.56,0,0,0,16.06,5c9.28,0,17-5.58,17-5.58,0-.71,19.57,11.74,25.33,14.35A27.41,27.41,0,0,1,118,109c2.37,4.53,5,16.43,5,16.43" fill="#202133" stroke="#000" stroke-miterlimit="10" stroke-width="0.75"/><path d="M44.34,48.38a7.48,7.48,0,0,0,14.95,0c0-4,1.12-7.22-7.47-7.22C44.81,41.16,44.34,44.39,44.34,48.38Z" fill="none" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><path d="M79.52,48.25a7.34,7.34,0,0,1-14.67,0c0-3.92-1.1-7.09,7.34-7.09C79.07,41.16,79.52,44.33,79.52,48.25Z" fill="none" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><path d="M65.14,47s-3.47-1.68-5.8,0" fill="none" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><path data-name="Collar" d="M47.05,83.4a6.53,6.53,0,0,0,.67-2,7.08,7.08,0,0,1,.7-2.5c.39-.43,1.64-.5,1.64-.5s-1,1.49-.46,2a24.37,24.37,0,0,0,3.9,3c1.58.93,8.39,4.45,9.13,4.79s-.09.51-.52.54-3.48.36-4.5,1.47-7.29,8.17-7.29,8.17S46.52,83.92,47.05,83.4Z" fill="#202133" stroke="#000" stroke-miterlimit="10" stroke-width="0.75"/><path data-name="collar" d="M76.73,78.09s.75-.34,1,.32.39,2,.45,2.32A22.2,22.2,0,0,0,81,83.18L78,97.57s-6.61-4.84-9.12-6.17-5.92-2.8-5.77-3,13-6.82,13.62-9.62a.81.81,0,0,0,0-.38Z" fill="#202133" stroke="#000" stroke-miterlimit="10" stroke-width="0.75"/><line x1="59.57" y1="88.99" x2="59.57" y2="125.47" fill="#fff" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><line x1="67.8" y1="90.52" x2="67.8" y2="125.47" fill="#fff" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><circle cx="63.68" cy="103.19" r="1.96" fill="#fff" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><circle data-name="button" cx="63.68" cy="119.16" r="1.96" fill="#fff" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><path d="M57.4,46.46c0,1.37-.56,2.48-1.26,2.48s-1.26-1.11-1.26-2.48,0-2.47,1.26-2.47C57,44,57.4,45.1,57.4,46.46Z" fill="#fff" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><path data-name="Eye" d="M73.63,46c0,1.47-.56,2.66-1.26,2.66S71.11,47.49,71.11,46s0-2.67,1.26-2.67C73.23,43.35,73.63,44.54,73.63,46Z" fill="#fff" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><path d="M57.17,50.31A3.8,3.8,0,0,1,55,51.37c-1.11.05-2.69-.57-3-1.35" fill="none" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><path data-name="undereye" d="M68.78,50.45a7,7,0,0,0,3,1.28,3.43,3.43,0,0,0,2.52-1.55" fill="none" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><path d="M65.18,49.86s2.07,8.13,2.15,8.85,1.16,3.37.83,3.78-.34,1.45-1.09,1.61A15.28,15.28,0,0,1,64.24,64a4.1,4.1,0,0,1-1.42-.5s-1.13.32-1.48.12a1.72,1.72,0,0,1-.74-1.28A5,5,0,0,1,61.51,61" fill="none" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><path d="M81,25.2l-.49,0c.32.24.53-.39.41-.76-1.62-4.94-6-8.67-10.92-10.35-5.25-1.8-12.83-2.29-17.63.85a24.85,24.85,0,0,0-9.77,12.34c-.6,1.66-1,3.38-1.75,5a7.74,7.74,0,0,1-3.42,3.88,4.23,4.23,0,0,1-4.87-.7A6,6,0,0,0,40,43.18l.14,3.11a6,6,0,0,1,3.2,3.26c1.11-1.94.76-4.43,1.73-6.45.82-1.71,2.57-3,2.7-4.92C48,35.4,44.52,33,51,29.42a1.11,1.11,0,0,1,.47-.14c2.19-.1,4.12,1.31,6.06,2.35a17.12,17.12,0,0,0,10.75,1.92,14.17,14.17,0,0,0,9.26-5.62,16.36,16.36,0,0,0-.67,7.24c.4,2.42,1.58,4.63,2.54,6.89a15.56,15.56,0,0,1,1.7,6.14.29.29,0,0,0,.55.16c.3-.51.52-1.06.79-1.61a14,14,0,0,0,.9-2.24,20.25,20.25,0,0,0-.7-13.43,1.59,1.59,0,0,1-.19-.79c.07-.68,1-.95,1.61-.64a3.17,3.17,0,0,1,1.23,1.57,4.49,4.49,0,0,0-4.32-6Z" fill="#7f500b" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/><path data-name="mouth" d="M54.58,64.56a4.57,4.57,0,0,1,2.36,1.78,8.5,8.5,0,0,0,4.4,2.94c2.45.63,3.6.84,3.35,2.1s-2.15,1.71-3.35,2.22a8,8,0,0,1-5.85-.95,4.92,4.92,0,0,1-2.24-3.46C52.8,67.39,53.14,63.82,54.58,64.56Z" fill="#ffffff" stroke="#020202" stroke-miterlimit="10" stroke-width="0.75"/></g></svg>',
  },\n  /** fs-placeholder */`
  )
)

// 生成pages
fs.outputFileSync(
  `./src/module/${name}/index.html`,
  `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>${cname}</title>
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="/src/module/${name}/app.ts"></script>
  </body>
</html>
`
)

// 生成 app.ts
fs.outputFileSync(
  `./src/module/${name}/app.ts`,
  `import { createApp } from 'vue'
import App from './app.vue'

import 'ant-design-vue/es/message/style/index'

const app = createApp(App)
app.mount('#app')\n`
)

// 生成 helper.ts
fs.outputFileSync(`./src/module/${name}/helper.ts`, `export {}\n`)

// 生成 helper.ts
fs.outputFileSync(
  `./src/module/${name}/interface.ts`,
  `export interface Example {
  label: string
  value: string
}\n`
)

// 生成 app.vue
fs.outputFileSync(
  `./src/module/${name}/app.vue`,
  `<script lang="ts" setup>
import type { Example } from './interface'

import { ref } from 'vue'
import { message } from 'ant-design-vue'

// 获取channel
import { useChannel } from '../../shared/use'
const { send: api__example } = useChannel('qmai.${name}.example')

const num = ref(1)

const onExample = async () => {
  const response = await api__example<number, Example>(num.value)
  message.info(JSON.stringify(response, null, 4))
}
</script>

<template>
  <br />

  <a-button type="primary" @click="onExample">测试</a-button>
</template>

<style>
body {
  background-color: #22272f !important;
}
</style>\n`
)

// 生成 extension.ts
fs.outputFileSync(
  `./src/module/${name}/extension.ts`,
  `import * as vscode from 'vscode'
import Channel from '@luozhu/vscode-channel'
import type { Example } from './interface'
import type { IResult } from '../../shared/interface'
import { activedPanelMap, createPanel } from '../../shared'
import { Result } from '../../shared/result'


const command = 'qmai.${name}'

export default (context: vscode.ExtensionContext) => {
  // 注册事件
  vscode.commands.registerCommand(command, () => {
    // 是否处于激活状态
    const isActive = activedPanelMap.has(command)

    // isTrue
    if (isActive) {
      // 重新聚焦
      activedPanelMap.get(command)?.reveal(vscode.ViewColumn.Active, true)
      return
    }

    const panel = createPanel({ command, title: '${cname}', pages: 'src/module/${name}/index.html', context })

    // 创建channel
    const channel = new Channel(context, panel)

    // 接收“示例”指令
    channel.bind<number, IResult<Example>>('qmai.${name}.example', (num) => {
      console.log(\`接收到了webview传递过来的数据 ==> \${num}\`)
      return Result.success({ label: '这里是label', value: '88888888' })
    })
  })
}\n`
)

const extensionContent = fs.readFileSync('./src/extension.ts', 'utf-8')
fs.writeFileSync(
  `./src/extension.ts`,
  extensionContent
    .replace('/** placeholder - 1 */', `import init${name} from './module/${name}/extension'\n/** placeholder - 1 */`)
    .replace('/* placeholder - 2 */', `init${name}(context)\n  /* placeholder - 2 */`)
)

await $`pnpm run build-view`
